core/primitives/src/sharding.rs

```diff
@@ -236,14 +616,101 @@ pub struct ReceiptList<'a>(pub ShardId, pub &'a Vec<Receipt>);
 struct TransactionReceipt(Vec<SignedTransaction>, Vec<Receipt>);
 
 #[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
-pub struct EncodedShardChunk {
+pub struct EncodedShardChunkV1 {
+    pub header: ShardChunkHeaderV1,
+    pub content: EncodedShardChunkBody,
+}
+
+impl EncodedShardChunkV1 {
+    pub fn chunk_hash(&self) -> ChunkHash {
+        self.header.chunk_hash()
+    }
+
+    pub fn decode_chunk(&self, data_parts: usize) -> Result<ShardChunkV1, std::io::Error> {
+        let transaction_receipts = EncodedShardChunk::create_transaction_receipts(
+            &self.content.parts[0..data_parts],
+            self.header.inner.encoded_length,
+        )?;
+
+        Ok(ShardChunkV1 {
+            chunk_hash: self.header.chunk_hash(),
+            header: self.header.clone(),
+            transactions: transaction_receipts.0,
+            receipts: transaction_receipts.1,
+        })
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+pub struct EncodedShardChunkV2 {
     pub header: ShardChunkHeader,
     pub content: EncodedShardChunkBody,
 }
 
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+pub enum EncodedShardChunk {
+    V1(EncodedShardChunkV1),
+    V2(EncodedShardChunkV2),
+}
+
 impl EncodedShardChunk {
-    pub fn from_header(header: ShardChunkHeader, total_parts: usize) -> Self {
-        Self { header, content: EncodedShardChunkBody { parts: vec![None; total_parts] } }
+    pub fn cloned_header(&self) -> ShardChunkHeader {
```

Comment:
maybe it's better to have `header()` which returns `&ShardChunkHeader` and caller can clone at call site?

Reply:
I really wanted to have a function like that, yes; it would have made many of the other code changes much cleaner. But the issue is the same as I described in the case of `ChunksCollection`. The `V1` variant needs to be lifted into the enum, but this requires creating new data in the local function call, which you cannot then return a reference to (because it would be released when the function call ends). I tried to find a way around this, but didn't find anything I really liked. If you know of a good solution please let me know.

Reply:
I see. I don't have a good solution either